tcmalloc/internal/percpu_rseq_aarch64.S

```diff
@@ -0,0 +1,487 @@
+ *   return len - r11
+ * }
+ */
+  .p2align 6 /* aligns to 2^6 with NOP filling */
+  .globl TcmallocSlab_PushBatch_FixedShift
+  .type  TcmallocSlab_PushBatch_FixedShift, @function
+TcmallocSlab_PushBatch_FixedShift:
+  .cfi_startproc
+.LTcmallocSlab_PushBatch_FixedShift_region0:
+  START_RSEQ(TcmallocSlab_PushBatch_FixedShift)
+  FETCH_CPU(w8)
+  lsl x8, x8, #PERCPU_TCMALLOC_FIXED_SLAB_SHIFT /* multiply cpu by 256k */
+  add x8, x0, x8
+  add x4, x8, x1, LSL #3
+  ldrh w9, [x4] /* current */
+  ldrh w10, [x4, #6] /* end */
+  cmp w9, w10
+  bge .LTcmallocSlab_PushBatch_FixedShift_region2
+  mov  x11, x3      /* r11 = copy of len */
+  sub  w10, w10, w9 /* r10 = free capacity */
+  cmp w11, w10
+  csel w10, w11, w10, ls /* r10 = min(len, free capacity) */
+  add x10, x10, x9
+.LTcmallocSlab_PushBatch_FixedShift_loop:
+  sub x11 ,x11, #1
+  ldr x12, [x2, x11, LSL #3]
+  str x12, [x8, x9, LSL #3]
+  add x9, x9, #1
+  cmp x9, x10
+  bne .LTcmallocSlab_PushBatch_FixedShift_loop
+  strh w9, [x4]
+.LTcmallocSlab_PushBatch_FixedShift_region1:
+.LTcmallocSlab_PushBatch_FixedShift_commit:
+  sub x0, x3, x11
+  ret
+.LTcmallocSlab_PushBatch_FixedShift_region2:
+  mov x0, #0
+  ret
+.LTcmallocSlab_PushBatch_FixedShift_region3:
+  .cfi_endproc
+ENCODE_SIZE(TcmallocSlab_PushBatch_FixedShift)
+DEFINE_UPSTREAM_CS(TcmallocSlab_PushBatch_FixedShift)
+
+/* size_t TcmallocSlab_PopBatch_FixedShift(
+ *     void *ptr (x0),
+ *     size_t cl (w1),
+ *     void** batch (x2),
+ *     size_t len (w3) {
+ *   uint64_t r8 = __rseq_abi.cpu_id
+ *   uint64_t* r8 = CpuMemoryStart(rdi, r8)
+ *   Header* hdr = GetHeader(rdi, rax, cl)
+ *   uint64_t r9 = hdr->current
+ *   uint64_t r10 = hdr->begin
+ *   if (r9 <= r10) return 0
+ *   r11 = min(rcx, r9 - r10)
+ *   rax = 0
+ * loop:
+ *   r9--
+ *   r10 = *(r8 + r9 * 8)
+ *   batch[rax] = r10
+ *   rax++
+ *   if (rax != r11) goto loop
+ *   hdr->current = r9
+ *   return rax
+ * }
+ */
+  .p2align 6 /* aligns to 2^6 with NOP filling */
+  .globl TcmallocSlab_PopBatch_FixedShift
+  .type  TcmallocSlab_PopBatch_FixedShift, @function
+TcmallocSlab_PopBatch_FixedShift:
+  .cfi_startproc
+.LTcmallocSlab_PopBatch_FixedShift_region0:
+  START_RSEQ(TcmallocSlab_PopBatch_FixedShift)
+  FETCH_CPU(w8)
+  lsl x8, x8, #PERCPU_TCMALLOC_FIXED_SLAB_SHIFT /* multiply cpu by 256k */
+  add x8, x0, x8
+  add x4, x8, x1, LSL #3
+  ldrh w9, [x4] /* current */
+  ldrh w10, [x4, #4] /* begin */
+  cmp w10, w9
+  bhs .LTcmallocSlab_PopBatch_FixedShift_region2
+  sub w11, w9, w10 /* w11 = available items */
+  cmp w3, w11
+  csel w11, w3, w11, ls /* r11 = min(len, available items) */
+  mov x12, #0
+.LTcmallocSlab_PopBatch_FixedShift_loop:
+  sub x9, x9, #1
+  ldr x10, [x8, x9, LSL #3]
+  str x10, [x2, x12, LSL #3]
+  add x12, x12, 1
+  cmp x12, x11
+  bne .LTcmallocSlab_PopBatch_FixedShift_loop
+  strh w9, [x4]
+.LTcmallocSlab_PopBatch_FixedShift_region1:
+.LTcmallocSlab_PopBatch_FixedShift_commit:
+  mov x0, x12
```

Comment:
`x0` isn't used in the loop, so I think you could rename `x12` to `x0` and eliminate this.

Reply:
x0 is used in the critical section though, so we can't overwrite it until we are out of it. Since if we abort this critical section a restart would require x0 to contain the original argument value. Or did I mis understand this?

Reply:
Ah, I didn't understand that requirement.